### PR TITLE
fix:(disk_setup):handle empty disk in check_partition_gpt_layout_sfdisk

### LINKED ISSUE #6682: Cloud-init fails to partition and empty disk with sfdisk when the type is gpt
# Bug report
When using RHEL10 as base image, only sfdisk is available and it fails to partition and empty disk when the table type is GPT. 

during cc_disk_setup, this command is run:

subp.py[DEBUG]: Running command ['sfdisk', '-l', '-J', '/dev/vdb'] with allowed return codes [0] (shell=False, capture=True)

However on an empty disk, the exit code is 1:

Error running partition command on /dev/vdb
Unexpected error while running command.
Command: ['sfdisk', '-l', '-J', '/dev/vdb']
Exit code: 1
Reason: -
Stdout: 
Stderr: sfdisk: /dev/vdb: does not contain a recognized partition table 
2026-01-22 13:21:13,952 - log_util.py[DEBUG]: Failed partitioning operation

## Steps to reproduce the problem
Cloud-init disk setup configured as:

   disk_setup:
      /dev/vdb:
        table_type: 'gpt'
        layout: True
        overwrite: true

## Environment details
- Cloud-init version: 24.4-6.el10
- Operating System Distribution: RHEL 10.1
- Cloud provider, platform or installer type: OCP

## cloud-init logs
2026-01-19 13:20:47,582 - cc_disk_setup.py[DEBUG]: Partitioning disks: {'/dev/vdb': {'layout': True, 'overwrite': True, 'table_type': 'gpt'}}
2026-01-19 13:20:47,582 - [subp.py](https://subp.py/)[DEBUG]: Running command ['udevadm', 'settle'] with allowed return codes [0] (shell=False, capture=True)
2026-01-19 13:20:47,590 - cc_disk_setup.py[DEBUG]: Checking values for /dev/vdb definition
2026-01-19 13:20:47,590 - cc_disk_setup.py[DEBUG]: Checking against default devices
2026-01-19 13:20:47,591 - cc_disk_setup.py[DEBUG]: Checking if device /dev/vdb is a valid device
2026-01-19 13:20:47,591 - [subp.py](https://subp.py/)[DEBUG]: Running command ['lsblk', '--pairs', '--output', 'NAME,TYPE,FSTYPE,LABEL', '/dev/vdb', '--nodeps'] with allowed return codes [0] (shell=False, capture=True)
2026-01-19 13:20:47,594 - cc_disk_setup.py[DEBUG]: Checking if device layout matches
2026-01-19 13:20:47,594 - [subp.py](https://subp.py/)[DEBUG]: Running command ['sfdisk', '-l', '-J', '/dev/vdb'] with allowed return codes [0] (shell=False, capture=True)
2026-01-19 13:20:47,604 - [performance.py](https://performance.py/)[DEBUG]: Creating partition on /dev/vdb took 0.022 seconds
2026-01-19 13:20:47,604 - log_util.py[WARNING]: Failed partitioning operation
Error running partition command on /dev/vdb
Unexpected error while running command.
Command: ['sfdisk', '-l', '-J', '/dev/vdb']
Exit code: 1
Reason: -
Stdout: 
Stderr: sfdisk: /dev/vdb: does not contain a recognized partition table
2026-01-19 13:20:47,604 - log_util.py[DEBUG]: Failed partitioning operation
Error running partition command on /dev/vdb
Unexpected error while running command.
Command: ['sfdisk', '-l', '-J', '/dev/vdb']
Exit code: 1
Reason: -
Stdout: 
Stderr: sfdisk: /dev/vdb: does not contain a recognized partition table
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_disk_setup.py", line 779, in check_partition_gpt_layout
    out, _err = subp.subp(prt_cmd, update_env=LANG_C_ENV)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/[subp.py](https://subp.py/)", line 291, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['sfdisk', '-l', '-J', '/dev/vdb']
Exit code: 1
Reason: -
Stdout: 
Stderr: sfdisk: /dev/vdb: does not contain a recognized partition table

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_disk_setup.py", line 58, in handle
    mkpart(disk, definition)
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_disk_setup.py", line 1103, in mkpart
    if check_partition_layout(table_type, device, layout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_disk_setup.py", line 813, in check_partition_layout
    found_layout = check_partition_gpt_layout(device, layout)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_disk_setup.py", line 787, in check_partition_gpt_layout
    raise RuntimeError(
RuntimeError: Error running partition command on /dev/vdb
Unexpected error while running command.
Command: ['sfdisk', '-l', '-J', '/dev/vdb']
Exit code: 1
Reason: -
Stdout: 
Stderr: sfdisk: /dev/vdb: does not contain a recognized partition table

### PR DESCRIPTION
**Pull request by xiachen-rh**
When checking GPT partition layout using sfdisk, handle the case where a disk has no partition table yet.
Previously, it would raise an exception.
Because check_partition_gpt_layout_sfdisk calls "sfdisk -l -J ", the command exits with code 1 and the error "does not contain a recognized partition table". The code doesn't handle it and raises an error.
Now it returns an empty list.
I also add unit test to cover this empty disk scenario.

Fixes GH-6682

## Test Steps
I run test on OpenStack with an additional empty disk with config data as below

```
# cat /etc/cloud/cloud.cfg.d/disk.cfg 
#cloud-config
disk_setup:
  /dev/vdb:
    table_type: 'gpt'
    layout: True
    overwrite: True
```
steps
   1) Create a volume and attach it to a instance
   2) login, and prepare configuration data for cloud-init to do disk setup
   3) clean cloud-init and reboot
   4) check the result

After fix, no unexpected error about sfdisk, and the additional disk /dev/vdb is parted
```
# lsblk 
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
vda    252:0    0   20G  0 disk 
├─vda1 252:1    0    1M  0 part 
├─vda2 252:2    0  200M  0 part /boot/efi
└─vda3 252:3    0 19.8G  0 part /
vdb    252:16   0   50G  0 disk 
└─vdb1 252:17   0   50G  0 part 
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by xiachen-rh** [comment]
Could anyone help review it please?

**Comment by xiachen-rh** [comment]
Hi @ani-sinha I would really appreciate it if you could review this code. Your feedback always helps me improve. Thanks!

**Review by ani-sinha — APPROVED** [review]
attached an empty disk to a loop device and ran the command manually. 

```
anisinha@virt:~$ dd if=/dev/zero of=blank.img bs=1M count=50
50+0 records in
50+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.0243454 s, 2.2 GB/s
anisinha@virt:~$ sudo mknod -m 660 /dev/loop0 b 7 0
anisinha@virt:~$ sudo chown root:disk /dev/loop0
anisinha@virt:~$ sudo losetup /dev/loop0 blank.img 
anisinha@virt:~$ sudo sfdisk -l -J /dev/loop0 2>/tmp/stderr.txt
anisinha@virt:~$ cat /tmp/stderr.txt 
sfdisk: /dev/loop0: does not contain a recognized partition table
```
seems I get the same error string on fc43. 
LGTM.

**Comment by xiachen-rh** [comment]
Hello @blackboxsw, this pull request will automatically close in 7 days, what can I do for it?

File: cloudinit/config/cc_disk_setup.py [hunk 1]
```diff
@@ -804,8 +804,14 @@ def check_partition_gpt_layout_sgdisk(device, layout):
 def check_partition_gpt_layout_sfdisk(device, layout):
     # Use sfdisk's JSON output for reliability
     prt_cmd = ["sfdisk", "-l", "-J", device]
+    _err = None
     try:
-        out, _err = subp.subp(prt_cmd, update_env=LANG_C_ENV)
+        out, _err = subp.subp(prt_cmd, update_env=LANG_C_ENV, rcs=[0, 1])
+        # Check if the error indicates no partition table exists
+        if _err and "does not contain a recognized partition table" in _err:
```
Comment (holmanb) [inline cloudinit/config/cc_disk_setup.py]: Depending on error strings like this is fragile. This code would no longer work correctly if sfdisk's error message changed, or if output is locale-specific then this wouldn't fix the issue for other locales.

Comment (xiachen-rh) [reply cloudinit/config/cc_disk_setup.py]: You’re right, thanks for pointing that out. Let me figure out how to fix it.

Comment (xiachen-rh) [reply cloudinit/config/cc_disk_setup.py]: Hello @holmanb , I resubmit the PR. What if we simply return [] when there’s no output? Given that rcs=1 may represent different failure cases, would returning [] be appropriate for all of them?

**Comment by xiachen-rh** [comment]
@blackboxsw  Could you please keep this PR open, I am waiting for review. thanks.

**Comment by ani-sinha** [comment]
cc: @holmanb @blackboxsw @TheRealFalcon

**Review by blackboxsw — APPROVED** [review]
Thank you for this contribution @xiachen-rh. I agree that treatment of an empty 'out' will avoid the intiial traceback seen in #6728 . 

That would allow the rest of cc_disk_setup to attempt to setup a device partition on a device.

The potential downfall of this approach could be that this empty list may hide other types exit 1 errors such as I/O errors or failures to dump partition table.

In those cases we will fall through to the partition layout setup request and deal 
with those types of errors if they persist during partition creation later.

**Comment by xiachen-rh** [comment]
Thanks @blackboxsw for the thoughtful comment, I agree with your concern about hiding other exit code 1 errors. I don’t have a better approach at the moment, but I’ll keep this in mind for future improvements.
